### ISSUE TITLE
Mass-Limited Guns and Turrets

### ISSUE THREAD

**Issue by Ferociousfeind** [endless-sky/endless-sky#9483]
### Problem Description

This is an Official(tm) depository for discussion of mass-limited hardpoints, since it was noticed in the ES discord that there actually wasn't any such issue on the github for ES.

For some time, content creators have wanted the ability to fine-tune the weapons which can fit into the mounts on a ship. Perhaps they want to restrict a certain class of weapon to only the largest of warships (and only the largest hardpoints on those warships), or perhaps they want to allow the player to invest an absurd amount of weapon capacity and hardpoint count on many small weapons, but also prevent the player from ignoring most hardpoints and investing all the weapon capacity into one big weapon.

### Related Issue Links

There are no related issues, is the main issue, really!

### Desired Solution

```
ship example
  attributes
    category "Heavy Warship"
    "outfit space" 800
    "weapon capacity" 250
    "engine capacity" 130
  "gun port" 0 -200
  "gun port" -20 -150
    limit 80
  "gun port" 20 -150
    limit 80
  "turret mount" 0 -50
  "turret mount" 0 50
    limit 50
```
This would allow a ship with one huge turret mount and one huge gun port, plus two okay-sized gun ports and one okay-sized turret mount. You could funnel all 250t of weapon capacity into the frontmost gun port (0, -200), or into the forward turret mount (0, -50), or you could divvy up the space fairly evenly between the limited ports. If they're fully capped, that leaves 40t of weapon space between the two unlimited-size ports.

This could be used fairly easily to allow something like the Protector to have a ton of hardpoints and a ton of space, but not allow the player to simply focus it _all_ into one of those hardpoints. (Did I say that already?) It could allow certain ships to have gun ports which have very low limits (only missile launchers and the like usable) to make them fairly obviously non-optimal for actual frontrunning military work, or to allow interceptors to have additional gun ports which are only usable for small additional damage output (you can mount a missile onto the tip of a plane, but you can't mount a particle cannon there).

The uses are admittedly not boundless, but the ability to prevent total unadulterated freedom from the player can be very useful in specific cases, _like the Emerald Sword._

### Alternative Approaches

Of course, the alternate solution is already in vanilla- a "spinal mount" attribute which is, by convention, provided by certain very large ships, and used up by certain very large guns (and turret). However, this has a major limiting factor, however, in that this only bars the large weapons from ships that don't happen to have this attribute, and it cannot distinguish between hardpoints- so these ships by convention only have one gun port (since all the spinal weapons are guns, except for the turret) to restrict the big gun to the one place it is supposed to go. (Needless to say, this is limiting and should be considered a temporary solution until a more robust system like this becomes implemented.)

**Comment by Hecter94** [comment]
Related issue, if not in execution, but in concept:
https://github.com/endless-sky/endless-sky/issues/5443

**Comment by Zitchas** [comment]
This has been a concept that has come up repeatedly in discussions over the years. It would add a lot of valuable limitations and detail to our ships. 

Going a step farther, I suspect that whenever we did this would probably be a good time to implemented a closely related  but slightly different feature: Require attributes in order to install a weapon in that particular hardpoint.

Actually, it might be better to just implement that, since mass (or outfit space, spinal attribute, weapon type, etc.) are all attributes, so simply implementing attribute based limitations on a hardpoint would enable mass/outfit space based limitations (which I would expect to be the most common use for it).

**Comment by OcelotWalrus** [comment]
> Going a step farther, I suspect that whenever we did this would probably be a good time to implemented a closely related but slightly different feature: Require attributes in order to install a weapon in that particular hardpoint.

So turret mounts and gun ports will have a new key where you'll list custom weapon attributes?

Something like this?

```
ship example
  attributes
    [...]
  "gun port" 0 -200
  "gun port" -20 -150
    attributes "attribute one"
  "gun port" 20 -150
    attributes "attribute one"
  "turret mount" 0 -50
  "turret mount" 0 50
    attributes "attribute one" "attribute two"
```
```
weapon example
  port attributes
    "attribute one"
  [...]
```
```
another weapon example
  port attributes
    "attribute two"
  [...]
```

**Comment by Zitchas** [comment]
Yes, that would work quite well. One of the big initial uses one could test with would be the spinal weapons, which are supposed to be locked to the one specific spinal slot.

**Comment by OcelotWalrus** [comment]
I think I'd be able to code a PR for that.

**Comment by Ferociousfeind** [comment]
You're right, although that would significantly slow down the ship-builfing process by taking the player out of the outfitter and into the info screen (or something similar) for every gun or turret they install.

Rather, (don't know if I said this already but) you can simply sort weapons onto their mounts in a particular order to guarantee that the ship will always have as much leniency for the player's whims as possible, and that whatever outfits on-paper should fit, will fit, if they can. (Specifically, whenever you install a weapon, simply put it into the smallest available mount, so that the weapon takes up the least excess resources, and what is left is the largest mounts for whatever else the player wants)

**Comment by 1010todd** [comment]
> You're right, although that would significantly slow down the ship-builfing process by taking the player out of the outfitter and into the info screen (or something similar) for every gun or turret they install.

The actual solution to that would be to put the hardpoint organizing part in the outfitter because right now requiring the player to leave the outfitter and go through several key presses is just  awful design. You should be able to do all your outfitting and ship customization in one place.

**Comment by AltinCloak** [comment]
For example, we already have a top down of the ship in the outfitter. You could just have a couple of selectable dot on it signifying different hardpoints. (Color coded, or they only show up when you have a gun or turret that fits there, including if you have custom attributes for the slot) and than the outfit goes into or out of that slot. They auto select in some order probably the order they are defined in the ship def and add a new child to a point definition that lets you offset the outfitter point off of its position. (I'm thinking for the bactrian and such.)

**Comment by Zitchas** [comment]
> The actual solution to that would be to put the hardpoint organizing part in the outfitter because right now requiring the player to leave the outfitter and go through several key presses is just awful design. You should be able to do all your outfitting and ship customization in one place.

This is definitely the best solution. I don't think it is necessary for this PR, but it is something that we should aim for. We already have the ship info screen, we just need to make that available in part of in whole inside the outfitter.
